cnd/src/btsieve/ethereum/ethereum_cache.rs

```diff
@@ -0,0 +1,142 @@
+use crate::{
+    btsieve::{
+        ethereum::{Block, Hash},
+        BlockByHash, LatestBlock, ReceiptByHash,
+    },
+    ethereum::TransactionReceipt,
+};
+use fmt::Debug;
+use futures::Future;
+use futures_core::{
+    compat::Future01CompatExt,
+    future::{FutureExt, TryFutureExt},
+};
+use std::{fmt, fmt::Formatter, sync::Arc};
+use tokio::sync::Mutex;
+
+pub struct EthereumCache<T> {
+    pub inner: T,
+    pub block_cache: Arc<Mutex<lru::LruCache<Hash, Option<Block>>>>,
+    pub receipt_cache: Arc<Mutex<lru::LruCache<Hash, Option<TransactionReceipt>>>>,
+}
+
+impl<T> Clone for EthereumCache<T>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        EthereumCache {
+            inner: self.inner.clone(),
+            block_cache: self.block_cache.clone(),
+            receipt_cache: self.receipt_cache.clone(),
+        }
+    }
+}
+
+impl<T> Debug for EthereumCache<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        self.inner.fmt(f)
+    }
+}
+
+impl<T> LatestBlock for EthereumCache<T>
+where
+    T: LatestBlock<Block = Option<Block>, BlockHash = Hash> + Clone,
+{
+    type Block = Option<Block>;
+    type BlockHash = Hash;
+
+    fn latest_block(
+        &mut self,
+    ) -> Box<dyn Future<Item = Self::Block, Error = anyhow::Error> + Send + 'static> {
+        let cache = self.block_cache.clone();
+        let mut inner = self.inner.clone();
+
+        let future = async move {
+            let block = inner.latest_block().compat().await?;
+
+            match block.clone() {
+                Some(block) => {
+                    let mut guard = cache.lock().await;
+                    guard.put(block.hash.expect("no blocks without hash"), Some(block));
+                }
+                None => (),
+            }
```

Comment:
Clippy suggests
```
            if let Some(block) = block.clone() {
                let mut guard = cache.lock().await;
                guard.put(block.hash.expect("no blocks without hash"), Some(block));
            };
```

Reply:
Also, should guard the `put` with a check for the blocks existence (same is suggested for bitcoin cache).